### ISSUE TITLE
Namespace inside modules

### ISSUE THREAD

**Issue by andyquinterom** [rstudio/py-shiny#307]
Hey there,

It might just be me, but how is someone supposed to access the namespace of the module from within the UI function? It might just be ignorance and lack of knowledge of python deeply, but on the R side, developing modules that depend heavily on external APIs, having the `ns <- NS(ID)` pattern is really useful.

Today I was livestreaming building a simple Chat App on py-shiny as a way to show the basics. I ran into some trouble with insert_ui inside modules. I wanted to insert into a div with an `id`, but be able to make it a re-usable module. In R, I would usually wrap the div's `id` inside the `ns` function and to the same on the server side. This worked perfect for me. However, I have no idea how it could be done in this version.

Is this posibble? Is this behaviour/pattern not desirable?

**Comment by jcheng5** [comment]
The idea in theory is that in Shiny for Python, you generally don't need to explicitly do `ns <- NS(id)` or `selectInput(inputId = ns(id))`; instead, inputs and outputs will automatically detect that they're in a module and namespace themselves automatically. However, I don't think we currently do that for regular div `id` attributes. Instead, you need to do it yourself for now. There are two ways to do this:

1. Use `resolve_id` (works in both module.ui and module.server)
```python
from shiny import module, ui

@module.ui
def mymodule_ui():
    return ui.div(id = module.resolve_id("container"))
```
2. Use `session.ns` (works only in module.server, since that's the only place you have `session`)
```python
@module.server
def mymodule_server(input, output, session):
    ui.insert_ui(ui.div(id = session.ns("container")), "body")
```

Please give this a try--I'm curious whether you find this better or worse than the R way, once you get used to it.

**Comment by andyquinterom** [comment]
Perfect. Would having id name spacing for divs be desirable. I feel like it could throw off some users, so it should be well documented.

If you feel it's all right I could work on a PR to implement such a feature

**Comment by Polkas** [comment]
I understand that namespaces are hidden in background now. 
Namespaces are one of the thing which was very hard to follow for newcomers so I think it is a great move. 

Creating a namespace is simply a concatenation of strings and is done only for protection against duplicated ids.
In my opinion adding namespaces is matter of courtesy, still I understand that shiny want o be sure that everything works correctly . It is like assuming sb could use "counter" id for both modules in the bottom example in https://shiny.rstudio.com/py/docs/workflow-modules.html.

In my opinion for sure ids should NOT be added to custom div automatically as this is an independent element where the full control has an end user, and such process is not needed for shiny to work securely. In the example above it is as simple as add a idx argument to the module ui and paste it to the core id.

**Comment by andyquinterom** [comment]
Sounds about right to me. We can close this issue.

**Comment by andyquinterom** [comment]
Perhaps we can open an issue about document this. It could have been helpful. I could work in some examples.

**Comment by jcheng5** [comment]
I think documentation is exactly the right answer. Looking now at the [modules](https://shiny.rstudio.com/py/docs/workflow-modules.html) doc they are really bare-bones! I'll file a doc issue for that (the repo is not currently public) and close this one if that's cool with everyone.
